app/scripts/controllers/user-storage/encryption.ts

```diff
@@ -23,14 +36,57 @@ function bytesToUtf8(byteArray: Uint8Array) {
   return decoder.decode(byteArray);
 }
 
+/**
+ * In Memory Cache of the derived key from a given salt.
+ * We can tidy this to make it reusable with future versions and KDF functions
+ */
+const inMemCachedKDF: Map<string, Uint8Array> = new Map();
+
+type CachedEntry = {
+  salt: Uint8Array;
+  key: Uint8Array;
+};
+const getCachedKeyBySalt = (salt: Uint8Array): CachedEntry | undefined => {
+  const base64Salt = byteArrayToBase64(salt);
+  const cachedKey = inMemCachedKDF.get(base64Salt);
+  if (!cachedKey) {
+    return undefined;
+  }
+
+  return {
+    salt,
+    key: cachedKey,
+  };
+};
+const getAnyCachedKey = (): CachedEntry | undefined => {
+  const cachedEntry = Array.from(inMemCachedKDF.entries()).at(0);
+  if (!cachedEntry) {
+    return undefined;
+  }
+
+  const bytesSalt = base64ToByteArray(cachedEntry[0]);
+  return {
+    salt: bytesSalt,
+    key: cachedEntry[1],
+  };
+};
+const setCachedKey = (salt: Uint8Array, key: Uint8Array): void => {
+  const base64Salt = byteArrayToBase64(salt);
+  inMemCachedKDF.set(base64Salt, key);
+};
+
 class EncryptorDecryptor {
   #ALGORITHM_NONCE_SIZE: number = 12; // 12 bytes
 
   #ALGORITHM_KEY_SIZE: number = 16; // 16 bytes
 
-  #PBKDF2_SALT_SIZE: number = 16; // 16 bytes
+  #SCRYPT_SALT_SIZE: number = 16; // 16 bytes
+
+  #SCRYPT_N: number = 2 ** 17; // CPU/memory cost parameter (must be a power of 2, > 1)
```

Comment:
This is the recommended value from OWASP. Another recommendation I saw elsewhere was 2^22 but I'm afraid that would be too big of a memory impact.

(should) Please do test the memory impact of current value (and maybe 22 too)
2^17 was documented to be 128MB of memory pressure, but with a JS implementation it might as well be 4x that if it's not implemented on raw byte arrays all the way.
Anyway, we want to know what the memory spike of this is. 

I recommend  "Allocation instrumentation on timeline" option in chromium devtools -> memory as the first thing to look at.

Reply:
Arbitrary JS testing on performance.

2**17 = ~1.5s
2**18 = ~1.8s
2**19 = ~2.4s
2**20 = ~4.4s
2**21 = ~7.5s
2**22 = ~13.8s

I'll also test memory performance, but >2s is a little bad for us (even if it is a 1 time cost).

If we want to support larger N values, we'll need to do a bigger refactor and keep a persisted cached key and create this key in the background.

To keep unblocked, is it possible to keep OWASP configuration and up this in a future PR?

Reply:
Memory timeframes show the allocations, but not to the max mem described in @noble readme. 

<details><summary>Memory recording (isolated through JEST)</summary>
Keep in mind these are isolated through JEST in case these are discrepancies. Will test on the actual application.

`2**17` results - peaked at 186KB
![Screenshot 2024-05-21 at 16 41 37](https://github.com/MetaMask/metamask-extension/assets/17534261/20b19cbe-76bc-436e-8fb7-083f8633e06f)


`2**22` results - peaked at 380KB
![Screenshot 2024-05-21 at 16 41 49](https://github.com/MetaMask/metamask-extension/assets/17534261/26d51e51-c455-43cb-b9a8-8afe7e4b3680)

</details>